src/core/api.c

```diff
@@ -1351,6 +1360,148 @@ MsQuicStreamReceiveComplete(
         "[ api] Exit");
 }
 
+_IRQL_requires_max_(DISPATCH_LEVEL)
+QUIC_STATUS
+QUIC_API
+MsQuicStreamProvideReceiveBuffers(
+    _In_ _Pre_defensive_ HQUIC Handle,
+    _In_ uint32_t BufferCount,
+    _In_reads_(BufferCount) const QUIC_BUFFER* Buffers
+    )
+{
+    QUIC_STATUS Status;
+    QUIC_OPERATION* Oper;
+    CXPLAT_LIST_ENTRY ChunkList;
+    CxPlatListInitializeHead(&ChunkList);
+
+    QuicTraceEvent(
+        ApiEnter,
+        "[ api] Enter %u (%p).",
+        QUIC_TRACE_API_STREAM_PROVIDE_RECEIVE_BUFFERS,
+        Handle);
+
+    if (!IS_STREAM_HANDLE(Handle) || Buffers == NULL || BufferCount == 0) {
+        Status = QUIC_STATUS_INVALID_PARAMETER;
+        goto Error;
+    }
+
+    for (uint32_t i = 0; i < BufferCount; ++i) {
+        if (Buffers[i].Length == 0) {
+            Status = QUIC_STATUS_INVALID_PARAMETER;
+            goto Error;
+        }
+    }
+
+#pragma prefast(suppress: __WARNING_25024, "Pointer cast already validated.")
+    QUIC_STREAM* Stream = (QUIC_STREAM*)Handle;
+
+    CXPLAT_TEL_ASSERT(!Stream->Flags.HandleClosed);
+    CXPLAT_TEL_ASSERT(!Stream->Flags.Freed);
+
+    QUIC_CONNECTION* Connection = Stream->Connection;
+    QUIC_CONN_VERIFY(Connection, !Connection->State.Freed);
+
+    //
+    // Execute this API call inline if called on the worker thread.
+    //
+    BOOLEAN IsWorkerThread = Connection->WorkerThreadID == CxPlatCurThreadID();
+    BOOLEAN IsAlreadyInline = Connection->State.InlineApiExecution;
+
+    if (!Stream->Flags.UseExternalRecvBuffers) {
+        if (Stream->Flags.PeerStreamStartEventActive) {
+            CXPLAT_DBG_ASSERT(IsWorkerThread);
+            //
+            // We are inline from the callback indicating a peer opened a stream.
+            // No data was received yet so we can setup external buffers.
+            //
+            Connection->State.InlineApiExecution = TRUE;
+            QuicStreamSwitchToExternalBuffers(Stream);
+            Connection->State.InlineApiExecution = IsAlreadyInline;
+        } else {
+            //
+            // External buffers can't be provided after the stream has been
+            // started using internal buffers.
+            //
+            Status = QUIC_STATUS_INVALID_STATE;
+            goto Error;
+        }
+    }
+
+    //
+    // Allocate a chunk for each buffer, linking them together.
+    // The allocation is done here to make the worker thread task failure free.
+    //
+    for (uint32_t i = 0; i < BufferCount; ++i) {
+        QUIC_RECV_CHUNK* Chunk = CXPLAT_ALLOC_NONPAGED(sizeof(QUIC_RECV_CHUNK), QUIC_POOL_RECVBUF);
```

Comment:
Should use a pool allocator (to be stored in the worker, like the others) since this is (a) fixed size and (b) on the datapath.

Reply:
This means each chunk will need a flag to track whether they are allocated from the pool or not then. The size of the chunk seems to be rather optimized (AllocLength restricted to 31bits to avoid a few extra bytes), is it ok to simply add a "flag" byte?

Which makes me realize I need to check the size of the app provided buffers won't override that bit.